components/masking_functions/src/masking_functions/random_string_generators.cpp

```diff
@@ -0,0 +1,214 @@
+/* Copyright (c) 2018, 2019 Francisco Miguel Biete Banon. All rights reserved.
+   Copyright (c) 2023 Percona LLC and/or its affiliates. All rights reserved.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software Foundation,
+   51 Franklin Street, Suite 500, Boston, MA 02110-1335 USA */
+
+#include "masking_functions/random_string_generators.hpp"
+
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#include <random>
+#include <string>
+#include <string_view>
+
+#include <boost/uuid/random_generator.hpp>
+#include <boost/uuid/uuid.hpp>
+#include <boost/uuid/uuid_io.hpp>
+
+namespace {
+
+std::default_random_engine &get_thread_local_prng() {
+  static thread_local std::default_random_engine instance{
+      std::random_device{}()};
```

Comment:
1. Investigated and verified that all UDFs are considered unsafe for SBR - there is nothing we can do about this. Moreover, these particular series of random functions is indeed unsafe.
2. The `thread_local` optimization was done primarily because of the `std::rundom_device` construction that is not a trivial operation and includes opening `/dev/urandom` as a file, reading data from it and closing the file handler. These three operations look like a huge overhead for each individual "random char" call. Ideally, one global `std::default_random_enginea  object per component would be enough but in this case we would need to protect it with some sync primitives. `thread_local` seems to me like a reasonable compromise.
3. In out case (on linux) `std::default_random_engine` is `minstd_rand` which is a `std::linear_congruential_engine<uint_fast32_t, 48271, 0, 2147483647>` - a Linear Congruential Engine `x[n+1] = (a*x[n] + c) mod m` with period about 2^31. It is not very hard to "hack" provided that you have several consecutive generated numbers. Potentially it can be substituted with `std::mersenne_twister_engine` which has a period of 2^19937 but to be honest I do not see a need for that. We are not using this generator for cryptography where it is vitally important to have a high quality RNG.

Added comments to `get_thread_local_prng()`